lib/node/runner/block_operations.go

```diff
@@ -47,58 +47,111 @@ func (sb *SavingBlockOperations) getNextBlock(height uint64) (nextBlock block.Bl
 
 func (sb *SavingBlockOperations) Check() (err error) {
 	sb.log.Debug("start to check")
-	return sb.check()
+
+	defer func() {
+		sb.log.Debug("finished to check")
+	}()
+
+	return sb.check(block.GetLatestBlock(sb.st).Height)
 }
 
 // continuousCheck will check the missing `BlockOperation`s continuously; if it
 // is failed, still try.
 func (sb *SavingBlockOperations) continuousCheck() {
 	for {
-		if err := sb.check(); err != nil {
+		if err := sb.check(sb.checkedBlock); err != nil {
 			sb.log.Error("failed to check", "error", err)
 		}
 
 		time.Sleep(5 * time.Second)
 	}
 }
 
-// check checks whether `BlockOperation`s of latest `block.Block` are saved; if
-// not it will try to catch up to the last.
-func (sb *SavingBlockOperations) check() (err error) {
-	var checked bool
-	var blk block.Block
-	for {
-		if blk, err = sb.getNextBlock(blk.Height); err != nil {
-			if err == errors.StorageRecordDoesNotExist {
-				if checked {
-					sb.log.Debug("stop checking; all the blocks are checked", "height", sb.checkedBlock)
-				}
-				err = nil
-			}
-
-			break
-		}
-
-		sb.log.Debug("check block", "block", blk.Hash)
-
-		var st *storage.LevelDBBackend
+func (sb *SavingBlockOperations) checkBlockWorker(id int, blocks <-chan block.Block, errChan chan<- error) {
+	var err error
+	var st *storage.LevelDBBackend
+	for blk := range blocks {
 		if st, err = sb.st.OpenBatch(); err != nil {
-			break
+			errChan <- err
+			return
 		}
 
 		if err = sb.CheckByBlock(st, blk); err != nil {
-			sb.log.Error("failed to check block", "block", blk.Hash, "height", blk.Height, "error", err)
 			st.Discard()
-			break
+			errChan <- err
+			return
 		}
+
 		if err = st.Commit(); err != nil {
 			sb.log.Error("failed to commit", "block", blk.Hash, "height", blk.Height, "error", err)
 			st.Discard()
-			break
+			errChan <- err
+			return
+		}
+
+		errChan <- nil
+	}
+}
+
+// check checks whether `BlockOperation`s of latest `block.Block` are saved; if
+// not it will try to catch up to the last.
+func (sb *SavingBlockOperations) check(lastBlock uint64) (err error) {
+	if lastBlock == common.GenesisBlockHeight {
+		return
+	}
+
+	blocks := make(chan block.Block)
+	errChan := make(chan error)
+	defer close(errChan)
+
+	numWorker := int(lastBlock / 2)
+	if numWorker > 100 {
+		numWorker = 100
+	} else if numWorker < 1 {
+		numWorker = 1
+	}
+
+	for i := 1; i <= numWorker; i++ {
+		go sb.checkBlockWorker(i, blocks, errChan)
+	}
+
+	go func() {
+		var blk block.Block
+		for {
+			if blk, err = sb.getNextBlock(blk.Height); err != nil {
+				err = errors.FailedToSaveBlockOperaton.Clone().SetData("error", err)
+				return
```

Comment:
Is it OK to not close the `blocks` here?

Reply:
@kfangw Oops, you right.